File: app/src/main/java/guillermo/com/pugme/interactors/PugsInteractorImpl.java
package guillermo.com.pugme.interactors;

import guillermo.com.pugme.callbacks.FetchPugsUrlsCallBack;
import guillermo.com.pugme.model.PugsModel;
import guillermo.com.pugme.model.retrofit.service.PugsService;
import retrofit.GsonConverterFactory;
import retrofit.Retrofit;
import retrofit.RxJavaCallAdapterFactory;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.schedulers.Schedulers;

/**
 * Created by guillermo.rosales on 21/01/16.
 */
public class PugsInteractorImpl implements PugsInteractor {

    private static int LIMIT = 50;
    private Retrofit retrofit = new Retrofit.Builder()
            .baseUrl(PugsService.BASE_URL)
            .addConverterFactory(GsonConverterFactory.create()).addCallAdapterFactory
                    (RxJavaCallAdapterFactory.create()).build();
    private PugsService.PugsServiceAPI service = retrofit.create(PugsService.PugsServiceAPI.class);

    @Override
    public void fetchPugsUrls(final FetchPugsUrlsCallBack callBack) {

        service.fetchImagesUrls(LIMIT).subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread()).subscribe(
                new Action1<PugsModel>() {
                    @Override
                    public void call(PugsModel sfDistrictsModels) {
                        callBack.onPugsUrlsFetched(sfDistrictsModels.getPugs());
                    }
                });

    }
}

File: app/src/main/java/guillermo/com/pugme/interactors/PugsInteractorImpl.kt
package guillermo.com.pugme.interactors

import guillermo.com.pugme.callbacks.FetchPugsUrlsCallBack
import guillermo.com.pugme.model.PugsModel
import guillermo.com.pugme.model.retrofit.service.PugsService
import retrofit.GsonConverterFactory
import retrofit.Retrofit
import retrofit.RxJavaCallAdapterFactory
import rx.android.schedulers.AndroidSchedulers
import rx.functions.Action1
import rx.schedulers.Schedulers


/**
 * Created by guillermo.rosales on 21/01/16.
 */
class PugsInteractorImpl : PugsInteractor {
    private val retrofit = Retrofit.Builder().baseUrl(PugsService.BASE_URL).addConverterFactory(GsonConverterFactory.create()).addCallAdapterFactory(RxJavaCallAdapterFactory.create()).build()
    private val service = retrofit.create<PugsService.PugsServiceAPI>(PugsService.PugsServiceAPI::class.java)

    override fun fetchPugsUrls(callBack: FetchPugsUrlsCallBack) {

        service.fetchImagesUrls(LIMIT).subscribeOn(Schedulers.newThread()).observeOn(AndroidSchedulers.mainThread()).subscribe { sfDistrictsModels -> callBack.onPugsUrlsFetched(sfDistrictsModels.pugs) }

    }

    companion object {

        private val LIMIT = 50
    }
}
